plugins/protocols/session-lock.cpp

```diff
@@ -0,0 +1,434 @@
+                LOGI("Focused ", output->name);
+            }
+
+            void handle_keyboard_leave(wf::seat_t *seat)
+            {
+                wlr_seat_keyboard_clear_focus(seat->seat);
+                LOGI("Unfocused ", output->name);
+            }
+
+            void handle_keyboard_key(wf::seat_t *seat, wlr_keyboard_key_event event)
+            {
+                wlr_seat_keyboard_notify_key(seat->seat, event.time_msec, event.keycode, event.state);
+            }
+
+          private:
+            wlr_surface *surface;
+            wlr_output *output;
+        };
+
+        wf::keyboard_interaction_t& keyboard_interaction()
+        {
+            return *interaction;
+        }
+
+      private:
+        wayfire_session_lock *lock;
+        wlr_session_lock_surface_v1 *lock_surface;
+        wf::output_t *output;
+        std::unique_ptr<wf::keyboard_interaction_t> interaction;
+        wf::wl_listener_wrapper lock_surface_destroy;
+    };
+
+    class backup_node_t : public simple_text_node_t
+    {
+      public:
+        backup_node_t(wf::output_t *output) : simple_text_node_t()
+        {
+            set_position({0, 0});
+            // TODO: it seems better to create the node and add it to the back of the scenegraph so
+            // it will be displayed if the client crashes and its surface is destroyed.
+            // Unfortunately this causes the surface to briefly appear before the lock screen.
+            // So make the background completely transparent instead, and then add the text
+            // when the client suraface is destroyed.
+            wf::cairo_text_t::params params(
+                1280 /* font_size */,
+                wf::color_t{0.1, 0.1, 0.1, 0} /* bg_color */,
+                wf::color_t{0.9, 0.9, 0.9, 1} /* fg_color */);
+            params.rounded_rect = false;
+            set_text_params(params);
+            set_size(output->get_screen_size());
+        }
+
+        void display()
+        {
+            wf::cairo_text_t::params params(
+                1280 /* font_size */,
+                wf::color_t{0, 0, 0, 1} /* bg_color */,
+                wf::color_t{0.9, 0.9, 0.9, 1} /* fg_color */);
+            set_text_params(params);
+            // TODO: make the text smaller and display a useful message instead of a big explosion.
+            set_text("💥");
+        }
+    };
+
+    struct output_state
+    {
+        output_state(wf::output_t *output) : output(output)
+        {}
+
+        wf::output_t *output;
+        std::shared_ptr<lock_surface_node_t> surface;
+        std::shared_ptr<backup_node_t> backup_surface;
+    };
+
+    class wayfire_session_lock
+    {
+      public:
+        wayfire_session_lock(wf_session_lock_plugin *plugin, wlr_session_lock_v1 *lock) :
+            plugin(plugin), lock(lock)
+        {
+            auto& ol = wf::get_core().output_layout;
+            output_added.set_callback([this] (wf::output_added_signal *ev)
```

Comment:
Instead of doing this manually, check out https://github.com/WayfireWM/wayfire/blob/master/src/api/wayfire/per-output-plugin.hpp#L30

You can add it as a base class and then it will connect to the signals, etc. for you.

Reply:
Looking at the code below, I assume you did this per hand simply because the mixin does not support non-plugin template types (yet). I think it should be very easy (and a good idea) to fix the mixin class - with an `if constexpr (std::is_base_v<per_output_plugin_instance_t, ConcretePluginType>) {}` around code blocks which are specific for per-output-plugin instances, so that the helper can be used generically.

Reply:
I did use the mixin for a while, to create the per-output objects. The reason I stopped using it is that the mixin expects the per-output instances to have zero-arg constructors. But here, the instances need to have a backpointer to a plugin-global object, for example to call notify_lock_state. Not sure if this is easy to do in the mixin - my command of C++ syntax isn't great.